platform/ext/target/nordic_nrf/common/core/native_drivers/spu.c

```diff
@@ -72,15 +72,23 @@ void spu_clear_events(void)
     nrf_spu_event_clear(NRF_SPU, NRF_SPU_EVENT_PERIPHACCERR);
 }
 
+bool spu_region_is_locked(NRF_SPU_Type * p_reg, uint8_t region_id)
+{
+    return p_reg->FLASHREGION[region_id].PERM & SPU_FLASHREGION_PERM_LOCK_Msk;
+}
+
 void spu_regions_reset_all_secure(void)
 {
     for (size_t i = 0; i < NUM_FLASH_SECURE_ATTRIBUTION_REGIONS ; i++) {
-        nrf_spu_flashregion_set(NRF_SPU, i,
-            1 /* Secure */,
-            NRF_SPU_MEM_PERM_READ
-            | NRF_SPU_MEM_PERM_WRITE
-            | NRF_SPU_MEM_PERM_EXECUTE,
-            0 /* No lock */);
+        if (!spu_region_is_locked(NRF_SPU, i))
+        {
+            nrf_spu_flashregion_set(NRF_SPU, i,
+                1 /* Secure */,
+                NRF_SPU_MEM_PERM_READ
+                | NRF_SPU_MEM_PERM_WRITE
+                | NRF_SPU_MEM_PERM_EXECUTE,
+                0 /* No lock */);
+            }
```

Comment:
I feel it is unfortunate that the function name states it is resetting all secure SPU regions while in actuality it is not.

I also feel it is dangerous for TF-M to assume that whatever was configured prior is appropriate.

Some kind of config enforcement, either build/config time or runtime is necessary I think.

Presumably TF-M is resetting all the config prior to doing some config. So it must know what it needs to configure and which regions need resetting.

Perhaps TF-M should reset the regions it intends to configure instead of resetting all regions?

Reply:
Also, patches to upstream projects need to be rationalized:

"This is not upstream because X. It can be removed when Y."

Reply:
SPU init essentially does:
* Set all regions as secure (this reset)
* Iterate through non-secure regions and set them up.
It does not specifically know which regions to configure as secure. If the region is locked TF-M cannot configure it to be different anyways (could fail with reset or something though).

I agree this is not nice, and it was also my concern when discussing this @Vge0rge earlier today.

Reply:
The TF-M build has accesss to pm_config.h I believe.

Perhaps it could infer from pm_config.h which SPU regions are assigned to the bootloaders and only skip configuring those regions if they are both locked and in a bootloader region?

Reply:
Thanks. I'll look into that